kapitan/inputs/cache.py

```diff
@@ -0,0 +1,80 @@
+import hashlib
+import logging
+import os
+import pickle
+from pathlib import Path
+from typing import Tuple
+
+from kapitan.errors import CompileError
+
+logger = logging.getLogger(__name__)
+
+
+class InputCache(object):
+    def __init__(self, input_type_name: str):
+        if cache_home := os.environ.get("XDG_CACHE_HOME"):
+            self.input_cache_home = f"{cache_home}/kapitan/{input_type_name}"
+        elif cache_home := os.environ.get("HOME"):
+            self.input_cache_home = f"{cache_home}/.cache/kapitan/{input_type_name}"
+        else:
+            raise CompileError("Could not get cache dir: $XDG_CACHE_HOME or $HOME not set.")
+
+        self.kv_cache = {}
+
+        logger.debug("Input cache home: %s", self.input_cache_home)
+
+    def hash_paths(self, inputs_hash) -> Tuple[Path, Path, Path]:
+        sub_path = Path(self.input_cache_home, inputs_hash[:2])
+        cached_path = Path(sub_path, inputs_hash[2:])
+        cached_path_lock = Path(str(cached_path) + ".lock")
+        return cached_path, cached_path_lock, sub_path
+
+    def get(self, inputs_hash, lock_retries=2) -> dict:  # output_obj
+        cached_path, cached_path_lock, _ = self.hash_paths(inputs_hash)
+        if not cached_path_lock.exists():
+            for retry in range(lock_retries):
+                try:
+                    with open(cached_path, "rb") as fp:
+                        logger.debug("Loading cache hit: %s", cached_path)
+                        return self.load_output(fp)
+                except FileNotFoundError:
+                    pass
+        return None
+
+    def set(self, inputs_hash, output_obj, lock_retries=2):
+        cached_path, cached_path_lock, sub_path = self.hash_paths(inputs_hash)
+        for retry in range(lock_retries):
+            # dont write if already exists
+            if cached_path.exists():
+                return inputs_hash
+
+            if not cached_path_lock.exists():
+                sub_path.mkdir(parents=True, exist_ok=True)
+                with open(cached_path_lock, "wb") as fp:
+                    logger.debug("Writing cache file: %s", cached_path_lock)
+                    self.dump_output(output_obj, fp)
+                cached_path_lock.rename(Path(str(cached_path_lock).removesuffix(".lock")))
```

Comment:
What if for some reason, someone aborts execution and this step is never executed? Should there be a recommendation to remove the `.lock`? Would it be to:
1. Delete the whole cache?
2. Delete just the `.lock` file?
3. Remove the `.locl` suffix?